File: src/mad2rando.py
import shutil
import os
import random
import logging
import configparser
import json

#print that the rando is working to a certain someone
print("Hey the randomizer is randomizing *cough* xeelium *cough*")

#Load/Create Config
configname = "rando.ini"
config = configparser.ConfigParser()
config['DEFAULT'] = {	'seed': '0',
						'sanity': 'True',
						'randomizeMinigames': 'False',
						'monkeylessLevels': 'True',
						'hasDeadends': 'False'}

if os.path.isfile(configname) == False:
	with open(configname, 'w') as configfile:
		config.write(configfile)
else:
	config.read(configname)

seed = int(config['DEFAULT']['seed'])
sanity = json.loads(config['DEFAULT']['sanity'].lower())
randomizeMinigames = json.loads(config['DEFAULT']['randomizeMinigames'].lower())
monkeylesslevels = json.loads(config['DEFAULT']['monkeylesslevels'].lower())
hasDeadends = json.loads(config['DEFAULT']['hasDeadends'].lower())

#Set basic vars
content = "Content"
contentog = "ContentOG"
streams = "/Streams/win/"
mptourney = "MultiplayerTourneyFinish"
bld = ".bld"
arc = ".arc"

#Setup the spoiler log
logger = logging.getLogger()
logger.setLevel(logging.INFO)
file_handler = logging.FileHandler('spoiler.log', mode='w')
file_handler.setLevel(logging.DEBUG)
logger.addHandler(file_handler)

#Level pools
deadend_pool = ["animal_chess", "Card_Match_Game", "Credits",\
"DivingLocation_IslandFever", "DivingLocation_PrepareToLaunch",\
"DivingLocation_RitesOfPassage", "DivingLocation_Waterhole",\
"DutyFree", "golf_cake", "golf_lovelylumps", "golf_targetTree",\
"HungryHippo", "Minigame_HotDurian", "RoP_MusicalChairs", "Soccer"]

minigame_levels = ["animal_chess", "Card_Match_Game",\
"DivingLocation_IslandFever", "DivingLocation_PrepareToLaunch",\
"DivingLocation_RitesOfPassage", "DivingLocation_Waterhole",\
"Minigame_Diving_Location_Menu", "HungryHippo", "Minigame_HotDurian",\
"RoP_MusicalChairs", "Soccer",\
"golf_3holes", "golf_baggagecheck",\
"golf_cratercrossing", "golf_crossedpaths", "golf_foosaBall",\
"golf_junkyard",  "golf_maze",\
"Golf_minigame", "golf_ravenousRhinos", "golf_cake",\
"golf_lovelylumps", "golf_targetTree"]

system_pool = ["map", "title", "global"]
required_pool = ["Dam_Busters", "FixThePlane", "Waterhole"]

#do nothing with these for now they are impossible to reach
impossible_boring_pool = ["penguins", "penguins2", "Prepare2Launch_Plane", "Prepare2Launch"]

campaign_pool = ["BraveNewWild", "ConvoyChase", "Credits",\
"DrMelman", "MartyRace", "RitesOfPassage",\
"VolcanoRave", "Watercaves", "Wooing_Gloria"]

boring_pool = ["DutyFree", "IslandFever", "Morts_Adventure"]
campaign_pool.extend(boring_pool)

minigame_pool = []

#Level blacklist doesn't allow these levels to be put into the final rando
blacklist = []

#Config handling
if hasDeadends == False:
	blacklist.extend(deadend_pool)

random.seed(seed)

if sanity == False:
	campaign_pool.extend(required_pool)	
else:
	system_pool.extend(required_pool)

if randomizeMinigames == True:
	campaign_pool.extend(minigame_levels)
else:
	minigame_pool.extend(minigame_levels)

#Lists of how many files we've yet to replace
levelscopied = []
levelscopied.extend(campaign_pool)
minigame_poolcopied = []
minigame_poolcopied.extend(minigame_pool)

#Finsh Config handling
if monkeylesslevels == True:
	campaign_pool.extend(impossible_boring_pool)
	levelscopied.extend(impossible_boring_pool)
else:
	blacklist.extend(boring_pool)

levelscopied.extend(boring_pool)

#Sort the pools so seeds are more stable
campaign_pool.sort()
levelscopied.sort()
minigame_pool.sort()
minigame_poolcopied.sort()

#Check for ContentOG
if os.path.isdir(contentog) == False:
	print("copy Content to ContentOG")
	quit()

#Replace content folder
shutil.rmtree(content)
os.makedirs(content+streams)

#Log the seed and the levels rando'd
logger.info('seed:'+str(seed))
logger.info('lvls:'+str(len(campaign_pool)))

#Blacklist lists setup
to_remove = []
to_add = []

#campaign pool blacklist filtering
for level in campaign_pool:
	if level in blacklist:
		to_remove.append(level)
		y = [item for item in campaign_pool if item not in blacklist]
		z = random.randrange(0, len(y))
		to_add.append(y[z])

campaign_pool = [item for item in campaign_pool if item not in to_remove]
campaign_pool.extend(to_add)
campaign_pool.sort()

#Blacklist lists cleanup
to_remove.clear()
to_add.clear()

#minigame pool blacklist filtering
if len(minigame_pool) != 0:
	for level in minigame_pool:
		if level in blacklist:
			to_remove.append(level)
			y = [item for item in minigame_pool if item not in blacklist]
			z = random.randrange(0, len(y))
			to_add.append(y[z])

minigame_pool = [item for item in minigame_pool if item not in to_remove]
minigame_pool.extend(to_add)
minigame_pool.sort()

#Copy campaign levels
for level in campaign_pool:
		x = random.randrange(0, len(levelscopied))
		shutil.copy(contentog+streams+level+arc, content+streams+levelscopied[x]+arc)
		shutil.copy(contentog+streams+level+bld, content+streams+levelscopied[x]+bld)
		logger.info('Main Pool Level:'+level+':'+levelscopied[x])
		levelscopied.pop(x)

#Copy minigame levels
if len(minigame_pool) != 0:
	for level in minigame_pool:
		x = random.randrange(0, len(minigame_poolcopied))
		shutil.copy(contentog+streams+level+arc, content+streams+minigame_poolcopied[x]+arc)
		shutil.copy(contentog+streams+level+bld, content+streams+minigame_poolcopied[x]+bld)
		logger.info('Minigame Pool Level:'+level+':'+minigame_poolcopied[x])
		minigame_poolcopied.pop(x)

#Copy system(unrandomized) levels
for level in system_pool:
	shutil.copy(contentog+streams+level+arc, content+streams+level+arc)
	shutil.copy(contentog+streams+level+bld, content+streams+level+bld)

shutil.copy(contentog+streams+mptourney+bld, content+streams+mptourney+bld)

#Say its done
print("Game randomized")

File: src/mad2rando.go
package main

import (
	//"fmt"
    "log"
    "os"
	//"github.com/bigkevmcd/go-configparser"
)

func main(){
	file, err := os.Create("spoiler.log")
	if err != nil {
        log.Fatal(err)
    }
	defer file.Close()

	SpoilerLogger := log.New(file, "SPOILER: ", 0)

	type Level struct {
		accessibleLevels []string
		hasCoins bool
		saneToRandomize bool
		monkeys int
		climbMonkeys int
		isMinigame bool
		breaksInMinigame bool
		boring bool
	}

	f := false;
	t := true;

	levels := map[string]Level{
		//accessible levels, hasCoins, saneToRandomize, monkeys, climbmonkeys, boring
		"animal_chess": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"Card_Match_Game": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"Credits": Level{[]string{"title", "map"}, f, t, 0, 0, f, f, f},
		"Minigame_Diving_Location_Menu": Level{[]string{"title", "DivingLocation_IslandFever", "DivingLocation_PrepareToLaunch", "DivingLocation_RitesOfPassage", "DivingLocation_Waterhole"}, f, t, 0, 0, t, f, f},
		"DivingLocation_IslandFever": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"DivingLocation_PrepareToLaunch": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"DivingLocation_RitesOfPassage": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"DivingLocation_Waterhole": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"DutyFree": Level{[]string{"title", "map"}, f, t, 0, 0, f, f, f},
		"golf_cake": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_lovelylumps": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_targetTree": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_3holes": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_baggagecheck": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_cratercrossing": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_crossedpaths": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_foosaBall": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_junkyard": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_maze": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"Golf_minigame": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"golf_ravenousRhinos": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"HungryHippo": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"Minigame_HotDurian": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"RoP_MusicalChairs": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"Soccer": Level{[]string{"title"}, f, t, 0, 0, t, f, f},
		"map": Level{[]string{"title", "DutyFree", "IslandFever"}, f, f, 0, 0, f, f, f},
		"title": Level{[]string{"DutyFree", "IslandFever"}, f, f, 0, 0, f, f, f},
		"Dam_Busters": Level{[]string{"Credits""title", "map"}, f, f, 0, 0, f, f, f},
		"FixThePlane": Level{[]string{"Waterhole", "Dam_Busters", "title", "map"}, f, f, 0, 0, f, t, f},
		"Waterhole": Level{[]string{"BraveNewWild", "FixThePlane", "DrMelman", "MartyRace", "RitesOfPassage", "VolcanoRave", "Watercaves", "ConvoyChase", "Morts_Adventure"}, f, f, 0, 0, f, f, f},
		"penguins": Level{[]string{"penguins2", "title", "map"}, f, t, 0, 0, f, f, f},
		"penguins2": Level{[]string{"ConvoyChase", "title", "map"}, f, t, 0, 0, f, f, f},
		"Prepare2Launch_Plane": Level{[]string{"BraveNewWild", "title", "map"}, f, t, 0, 0, f, f, f},
		"Prepare2Launch": Level{[]string{"Prepare2Launch_Plane", "title", "map"}, f, t, 0, 0, f, f, f},
		"BraveNewWild": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
		"ConvoyChase": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
		"DrMelman": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
		"MartyRace": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
		"RitesOfPassage": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
		"VolcanoRave": Level{[]string{"title", "map"}, f, t, 0, 0, f, f, f},
		"Watercaves": Level{[]string{"Waterhole", "Wooing_Gloria", "title", "map"}, f, t, 0, 0, f, f, f},
		"Wooing_Gloria": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
		"IslandFever": Level{[]string{"Prepare2Launch", "title", "map"}, f, t, 0, 0, f, f, f},
		"Morts_Adventure": Level{[]string{"Waterhole", "title", "map"}, f, t, 0, 0, f, f, f},
	}

	type LevelSlot struct {
		reqCoins bool
		reqMonkeys bool
		reqClimb bool
		isMinigame bool
		reqWooingGloria bool
	}

	levelslots := map[string]LevelSlot{
		"animal_chess": LevelSlot{f, f, f, t, f},
		"Card_Match_Game": LevelSlot{f, f, f, t, f},
		"Credits": LevelSlot{f, f, f, f, f},
		"Minigame_Diving_Location_Menu": LevelSlot{f, f, f, t, f},
		"DivingLocation_IslandFever": LevelSlot{f, f, f, t, f},
		"DivingLocation_PrepareToLaunch": LevelSlot{t, f, f, t, f},
		"DivingLocation_RitesOfPassage": LevelSlot{t, f, f, t, f},
		"DivingLocation_Waterhole": LevelSlot{f, f, f, t, f},
		"DutyFree": LevelSlot{f, f, f, t, f},
		"golf_cake": LevelSlot{t, f, f, t, f},
		"golf_lovelylumps": LevelSlot{t, f, f, t, f},
		"golf_3holes": LevelSlot{f, f, f, t, f},
		"golf_baggagecheck": LevelSlot{f, f, f, t, f},
		"golf_cratercrossing": LevelSlot{f, f, f, t, f},
		"golf_crossedpaths": LevelSlot{f, f, f, t, f},
		"golf_foosaBall": LevelSlot{f, f, f, t, f},
		"golf_junkyard": LevelSlot{f, f, f, t, f},
		"golf_maze": LevelSlot{f, f, f, t, f},
		"Golf_minigame": LevelSlot{f, f, f, t, f},
		"golf_ravenousRhinos": LevelSlot{f, f, f, t, f},
		"HungryHippo": LevelSlot{f, f, f, t, f},
		"Minigame_HotDurian": LevelSlot{f, f, f, t, f},
		"RoP_MusicalChairs": LevelSlot{f, f, f, t, f},
		"Soccer": LevelSlot{f, f, f, t, f},
		"map": LevelSlot{f, f, f, f, f},
		"title": LevelSlot{f, f, f, f, f},
		"Dam_Busters": LevelSlot{f, t, f, f, f},
		"FixThePlane": LevelSlot{f, f, f, f, f},
		"Waterhole": LevelSlot{f, f, f, f, f},
		"penguins": LevelSlot{f, f, f, f, f},
		"penguins2": LevelSlot{f, f, f, f, f},
		"Prepare2Launch_Plane": LevelSlot{f, f, f, f, f},
		"Prepare2Launch": LevelSlot{f, f, f, f, f},
		"BraveNewWild": LevelSlot{f, f, f, f, f},
		"ConvoyChase": LevelSlot{f, f, f, f, f},
		"DrMelman": LevelSlot{f, f, f, f, f},
		"MartyRace": LevelSlot{f, f, f, f, f},
		"RitesOfPassage": LevelSlot{f, f, f, f, f},
		"VolcanoRave": LevelSlot{f, f, f, f, t},
		"Watercaves": LevelSlot{f, f, t, f, f},
		"Wooing_Gloria": LevelSlot{f, f, f, f, f},
		"IslandFever": LevelSlot{f, f, f, f, f},
		"Morts_Adventure": LevelSlot{f, f, f, f, f},
	}

	//deadend_pool := []string{
	//	"animal_chess", "Card_Match_Game", "Credits",
	//	"DivingLocation_IslandFever", "DivingLocation_PrepareToLaunch",
	//	"DivingLocation_RitesOfPassage", "DivingLocation_Waterhole",
	//	"DutyFree", "golf_cake", "golf_lovelylumps", "golf_targetTree",
	//	"HungryHippo", "Minigame_HotDurian", "RoP_MusicalChairs", "Soccer",
	//}

	//minigame_levels := []string{
	//	"animal_chess", "Card_Match_Game",
	//	"DivingLocation_IslandFever", "DivingLocation_PrepareToLaunch",
	//	"DivingLocation_RitesOfPassage", "DivingLocation_Waterhole",
	//	"Minigame_Diving_Location_Menu", "HungryHippo", "Minigame_HotDurian",
	//	"RoP_MusicalChairs", "Soccer",
	//	"golf_3holes", "golf_baggagecheck",
	//	"golf_cratercrossing", "golf_crossedpaths", "golf_foosaBall",
	//	"golf_junkyard",  "golf_maze",
	//	"Golf_minigame", "golf_ravenousRhinos", "golf_cake",
	//	"golf_lovelylumps", "golf_targetTree",
	//}

	//SpoilerLogger.Println(levels["animal_chess"])

	for k, v := range levels {
		SpoilerLogger.Println(k, v)
	}
	
}